File: src/PoMiniGames.Client/src/games/tictactoe/TicTacToeBoard.ts
import { CellValue } from '../shared/types';

/** 6×6 Tic Tac Toe board with WinLength = 4. Immutable — every mutation returns a new board. */
export class TicTacToeBoard {
  static readonly Size = 6;
  static readonly WinLength = 4;

  private readonly cells: CellValue[][];

  constructor(cells?: CellValue[][]) {
    if (cells) {
      this.cells = cells;
    } else {
      this.cells = Array.from({ length: TicTacToeBoard.Size }, () =>
        Array.from({ length: TicTacToeBoard.Size }, () => CellValue.None),
      );
    }
  }

  /** Get cell value at (row, col). */
  get(row: number, col: number): CellValue {
    return this.cells[row]![col]!;
  }

  /** Return a new board with the piece placed at (row, col). */
  place(row: number, col: number, value: CellValue): TicTacToeBoard {
    if (this.cells[row]![col] !== CellValue.None) {
      throw new Error(`Cell (${row}, ${col}) is already occupied`);
    }
    const clone = this.cloneCells();
    clone[row]![col] = value;
    return new TicTacToeBoard(clone);
  }

  /** Check if the given player has won. Returns winning cells if so. */
  checkWin(player: CellValue): { won: boolean; cells: [number, number][] } {
    const S = TicTacToeBoard.Size;
    const W = TicTacToeBoard.WinLength;
    const directions = [
      [0, 1],   // horizontal
      [1, 0],   // vertical
      [1, 1],   // diagonal down
      [1, -1],  // diagonal up
    ] as const;

    for (let r = 0; r < S; r++) {
      for (let c = 0; c < S; c++) {
        for (const [dr, dc] of directions) {
          const line: [number, number][] = [];
          let valid = true;
          for (let i = 0; i < W; i++) {
            const nr = r + dr * i;
            const nc = c + dc * i;
            if (nr < 0 || nr >= S || nc < 0 || nc >= S || this.cells[nr]![nc] !== player) {
              valid = false;
              break;
            }
            line.push([nr, nc]);
          }
          if (valid) {
            return { won: true, cells: line };
          }
        }
      }
    }
    return { won: false, cells: [] };
  }

  /** True when every cell is occupied. */
  isFull(): boolean {
    for (let r = 0; r < TicTacToeBoard.Size; r++) {
      for (let c = 0; c < TicTacToeBoard.Size; c++) {
        if (this.cells[r]![c] === CellValue.None) return false;
      }
    }
    return true;
  }

  /** All empty cells as (row, col) tuples. */
  getAvailableMoves(): [number, number][] {
    const moves: [number, number][] = [];
    for (let r = 0; r < TicTacToeBoard.Size; r++) {
      for (let c = 0; c < TicTacToeBoard.Size; c++) {
        if (this.cells[r]![c] === CellValue.None) {
          moves.push([r, c]);
        }
      }
    }
    return moves;
  }

  private cloneCells(): CellValue[][] {
    return this.cells.map((row) => [...row]);
  }
}

File: src/PoMiniGames.Client/Games/TicTacToe/TicTacToeBoard.cs
using System.Runtime.InteropServices;
using PoMiniGames.Shared.Games;
using PoMiniGamesClient.Models;

namespace PoMiniGamesClient.Games.TicTacToe;

public class TicTacToeBoard
{
    // Geometry and the win check come from the shared rules so the online match
    // service (which applies moves through the same class) can never disagree
    // with what this board draws. CellValue is byte-backed for exactly this handoff.
    public const int Size = TicTacToeRules.BoardSize;
    public const int WinLength = TicTacToeRules.BoardWinLength;

    // Flat row-major storage (2026-09-14, online mode) — the jagged CellValue[][]
    // it replaced could not be viewed as the byte span the shared rules read.
    private readonly CellValue[] _cells;

    public TicTacToeBoard()
    {
        _cells = new CellValue[Size * Size];
    }

    private TicTacToeBoard(CellValue[] cells)
    {
        _cells = cells;
    }

    /// <summary>
    /// Rebuild a board from an authoritative flat cell array (the online match
    /// state). Used when the local board is more than one move behind the server —
    /// a rejoin, or a missed broadcast — so there is no placement to animate anyway.
    /// </summary>
    public static TicTacToeBoard FromCells(ReadOnlySpan<byte> cells)
    {
        if (cells.Length != Size * Size)
        {
            throw new ArgumentException($"Expected {Size * Size} cells, got {cells.Length}.", nameof(cells));
        }
        var values = new CellValue[Size * Size];
        MemoryMarshal.Cast<byte, CellValue>(cells).CopyTo(values);
        return new TicTacToeBoard(values);
    }

    public CellValue Get(int row, int col) => _cells[row * Size + col];

    public TicTacToeBoard Place(int row, int col, CellValue value)
    {
        if (_cells[row * Size + col] != CellValue.None)
        {
            throw new Exception($"Cell ({row}, {col}) is already occupied");
        }

        var cells = new CellValue[_cells.Length];
        Array.Copy(_cells, cells, _cells.Length);
        cells[row * Size + col] = value;
        return new TicTacToeBoard(cells);
    }

    public WinResult CheckWin(CellValue player)
    {
        var line = TicTacToeRules.Instance.FindWin(MemoryMarshal.AsBytes<CellValue>(_cells), (byte)player);
        if (line is null) return new WinResult { Won = false, Cells = new List<(int, int)>() };
        var cells = new List<(int, int)>(WinLength);
        foreach (var index in line)
        {
            cells.Add((index / Size, index % Size));
        }
        return new WinResult { Won = true, Cells = cells };
    }

    public bool IsFull() => TicTacToeRules.Instance.IsFull(MemoryMarshal.AsBytes<CellValue>(_cells));

    public List<(int Row, int Col)> GetAvailableMoves()
    {
        var moves = new List<(int, int)>();
        for (int i = 0; i < _cells.Length; i++)
        {
            if (_cells[i] == CellValue.None)
            {
                moves.Add((i / Size, i % Size));
            }
        }
        return moves;
    }
}
